### PR TITLE
Array parameters

### LINKED ISSUE #103: Add the possibility to pass an array to all `keyValue` actions
## Problem
We often have to update every values in a `keyValue` store. But with the current API, it can lead to a huge number of dispatched actions.

## Proposition
I propose to make `update`, `addOrUpdate` and `replace` action polymorphic on their argument, allowing to pass an instance or an array of instances.

## Example
```es6
import { keyValue } from 'k-redux-factory'

const store = keyValue({ key: 'id' })

store.addOrUpdate([
  { id: 1, value: '1' },
  { id: 2, value: '2' },
  { id: 3, value: '3' },
])
```

### PR DESCRIPTION
**Pull request by EmrysMyrddin**
fixes #103

### CODE REVIEW

File: src/types/keyValue/keyValue.middleware.js [hunk 1]
```diff
@@ -63,16 +63,40 @@ const reducer = key => prefix => name => defaultData =>
   (state = defaultState(key, defaultData), { type, payload } = {}) => {
     switch (type) {
       case SET(prefix)(name): return set(key, payload)
-      case ADD(prefix)(name): return add(key, state, payload)
+      case ADD(prefix)(name): {
+      if (Array.isArray(payload))
```
Comment (fabienjuif) [inline src/types/keyValue/keyValue.middleware.js]: I think you can run `yarn ci` and you'll see errors there

Comment (fabienjuif) [reply src/types/keyValue/keyValue.middleware.js]: I just activate the fork building into circleci

File: src/types/keyValue/keyValue.middleware.js [hunk 2]
```diff
@@ -63,16 +63,40 @@ const reducer = key => prefix => name => defaultData =>
   (state = defaultState(key, defaultData), { type, payload } = {}) => {
```
Comment (fabienjuif) [inline src/types/keyValue/keyValue.middleware.js]: Right now I have the feeling we can do an other function.

- **main**: default behaviour and calls the second fonction base on the parameter type
- **reduceWithArray**: for array type
- **reduceWithSingleIntance**: for single instance parameter

I have the feeling the code would be prettier to read

Comment (fabienjuif) [reply src/types/keyValue/keyValue.middleware.js]: Or factorizing this :
```js
(newState, entity) =>
            keyAlreadyExists(state)(key, entty[key])
              ? update(key, state, entity)
              : add(key, newState, entity)
```

```js
const applyReduce = (state, key, entty) => (existCallback, notExistCallback) =>  (newState, entity) =>
            keyAlreadyExists(state)(key, entty[key])
              ? existCallback(key, state, entity)
              : notExistCallback(key, newState, entity),

// ...
return payload.reduce(applyReduce(state, key, entty)(update, add))

```

**Comment by fabienjuif** [comment]
We should add tests

**Comment by EmrysMyrddin** [comment]
Yes, I haven't done the tests yet because I wasn't really satisfied of my implementation.
I will try your proposition.

But I'm afraid of some perfomance issues here. Why do we not use `immutadot` or an other library that allow to do immutable updates for us with pretty good perfs ?

**Comment by fabienjuif** [comment]
Because the purposes of this lib are:
 - Avoiding the **copy/paste of dumbs reducers**
 - Avoiding to have a growing useless **bundle size**
 - Focusing on **reading performance**

You can benchmark others libs to include:
 - **writing** performance
 - **bundle size**

But I bet this is not worth

**Comment by fabienjuif** [comment]
IMO, the performance issue could be there:
```js
const keyAlreadyExists =
  state => (key, instanceKey) => state.array.find(o => o[key] === instanceKey)
```

this function should be something like: 
```js
const keyAlreadyExists = state => (key) => state.keys[key] !== undefined
```

but this is an other story/PR

**Pull request closed without merging**